sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -770,21 +812,112 @@ static int create_sdw_codec_dai(struct device *dev,
 		if (!name)
 			return -ENOMEM;
 
-		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
-					  "SDW%d Pin%d", link_id, j + 2);
-		if (!cpu_name)
-			return -ENOMEM;
+		for (k = 0; k < cpu_num; k++) {
+			int index = SDW_DISCO_LINK_ID(link->adr[k]);
 
-		init_dai_link(dai_links, *id, name, 1 - i, i, cpus, cpu_name,
+			cpu_name = devm_kasprintf(dev, GFP_KERNEL,
+						  "SDW%d Pin%d", index, j + 2);
+			if (!cpu_name)
+				return -ENOMEM;
+
+			cpu_dai[k].dai_name = cpu_name;
 			      codec, link->num_adr, *id, NULL, &sdw_ops);
+
+		init_dai_link(dai_links, *id, name, 1 - i, i, cpu_dai, cpu_num,
 		set_codec_init_func(link, dai_links + *id, 1 - i);
+		cpu_dai += cpu_num;
 		(*id)++;
 		j++;
 	}
 
 	return 0;
 }
 
+static inline bool is_spk(int id)
+{
+	if (id == 0x1308)
+		return true;
+
+	return false;
+}
+
+static int convert_sdw_link_to_sdw_dai(struct device *dev,
+				       struct snd_soc_acpi_mach *mach,
+				       struct sdw_dai_link **sdw_dai,
+				       int *sdw_dai_num)
+{
+	const struct snd_soc_acpi_link_adr *adr_link;
+	const struct snd_soc_acpi_link_adr *link;
+	struct sdw_dai_link *tmp_link;
+	int spk_num, link_num;
+	u64 adr_data[4];
+	bool merge_spk;
+	int part_id;
+	int i, j;
+
+	merge_spk = sof_rt711_rt1308_rt715_quirk & SOF_SDW_MERGE_SPK_DAI;
+
+	adr_link = mach->mach_params.links;
+	link = adr_link;
+	for (link_num = 0, spk_num = 0; link->num_adr; link++) {
+		part_id = SDW_PART_ID(link->adr[0]);
+		if (is_spk(part_id))
+			adr_data[spk_num++] = link->adr[0];
+
+		link_num++;
+	}
+
+	if (merge_spk)
+		link_num = link_num - spk_num + 1;
+
+	tmp_link = devm_kcalloc(dev, link_num, sizeof(*tmp_link),
+				GFP_KERNEL);
+	if (!tmp_link)
+		return -ENOMEM;
+
+	for (i = 0, j = 0; adr_link->num_adr && (j < link_num); i++) {
+		const u64 *src;
+		int adr_size;
+		u64 *adr;
+
+		tmp_link[j].dai_id = ffs(adr_link->mask) - 1;
+		tmp_link[j].num_adr = adr_link->num_adr;
+
+		part_id = SDW_PART_ID(adr_link->adr[0]);
+		if (!merge_spk || !is_spk(part_id)) {
+			src = adr_link->adr;
+			adr_size = adr_link->num_adr * sizeof(u64);
+		} else {
+			src = adr_data;
+			adr_size = spk_num * sizeof(u64);
+			tmp_link[j].num_adr = spk_num;
+		}
+
+		adr = devm_kzalloc(dev, adr_size, GFP_KERNEL);
+		if (!adr)
+			return -ENOMEM;
+
+		memcpy(adr, src, adr_size);
+		tmp_link[j].adr = adr;
+
+		adr_link++;
+		if (!merge_spk || !is_spk(part_id)) {
+			j++;
+			continue;
+		}
+
+		if (spk_num) {
+			spk_num = 0;
+			j++;
+		}
+	}
+
+	*sdw_dai = tmp_link;
+	*sdw_dai_num = link_num;
+
+	return 0;
+}
```

Comment:
sorry @RanderWang there's not a single comment and this looks like spaghetti code. please explain what this is supposed to do, there are just too many variables changing here.